nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java

```diff
@@ -0,0 +1,243 @@
+        FlowFile flowFile = session.get();
+        if ( flowFile == null ) {
+            return;
+        }
+
+        JsonObjectBuilder builder = Json.createObjectBuilder();
+        String text = context.getProperty(WEBHOOK_TEXT).evaluateAttributeExpressions(flowFile).getValue();
+        if (text != null && !text.isEmpty()) {
+            builder.add("text", text);
+        } else {
+            // Slack requires the 'text' attribute
+            getLogger().error("FlowFile should have non-empty " + WEBHOOK_TEXT.getName());
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        String channel = context.getProperty(CHANNEL).evaluateAttributeExpressions(flowFile).getValue();
+        if (channel != null && !channel.isEmpty()) {
+            String error = validateChannel(channel);
+            if (error == null) {
+                builder.add("channel", channel);
+            } else {
+                getLogger().error("Invalid channel '{}': {}", new Object[]{channel, error});
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+        }
+
+        String username = context.getProperty(USERNAME).evaluateAttributeExpressions(flowFile).getValue();
+        if (username != null && !username.isEmpty()) {
+            builder.add("username", username);
+        }
+
+        String iconUrl = context.getProperty(ICON_URL).evaluateAttributeExpressions(flowFile).getValue();
+        if (iconUrl != null && !iconUrl.isEmpty()) {
+            builder.add("icon_url", iconUrl);
+        }
+
+        String iconEmoji = context.getProperty(ICON_EMOJI).evaluateAttributeExpressions(flowFile).getValue();
+        if (iconEmoji != null && !iconEmoji.isEmpty()) {
+            builder.add("icon_emoji", iconEmoji);
+        }
+
+        JsonObject jsonObject = builder.build();
+        StringWriter stringWriter = new StringWriter();
+        JsonWriter jsonWriter = Json.createWriter(stringWriter);
+        jsonWriter.writeObject(jsonObject);
+        jsonWriter.close();
+
+        try {
+            URL url = new URL(context.getProperty(WEBHOOK_URL).getValue());
+            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+            conn.setRequestMethod("POST");
+            conn.setDoOutput(true);
+            DataOutputStream outputStream  = new DataOutputStream(conn.getOutputStream());
+            String payload = "payload=" + URLEncoder.encode(stringWriter.getBuffer().toString(), "UTF-8");
+            outputStream.writeBytes(payload);
+            outputStream.close();
+
+            int responseCode = conn.getResponseCode();
+            if (responseCode >= 200 && responseCode < 300) {
+                getLogger().info("Successfully posted message to Slack");
+                session.transfer(flowFile, REL_SUCCESS);
```

Comment:
I thought about using `url.toString()`, but there is a lot of <a href="http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/net/URL.java#503">non-trivial parsing going on inside `URL`</a> which made it less appealing.